### PR TITLE
Discrete geometric pricer under Heston vol

### PR DESCRIPTION
**Pull request by jackgillett101**
Pricing discrete geometric asians under Heston volatility, using the method described by B. Kim, J. Kim, J. Kim & I. S. Wee in "A Recursive Method for Discretely Monitored Geometric Asian Option Prices"

### CODE REVIEW

**Comment by jackgillett101** [comment]
Impacts https://github.com/google/tf-quant-finance/issues/56

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+
+        rows.append(row)
+        last_row = row
+
+    rows = list(reversed(rows))
+    omega_tilde = tf.stack(rows)
+
+    return omega_tilde
+
+  def Phi(self, s, w):
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:482]: nit: we'd need to follow style guide  and call it `phi`

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+    x = tf.math.sqrt(tf.math.pow(self.mean_reversion, 2)
+                     - 2 * z1 * tf.math.pow(self.volvol, 2))
+
+    term_1 = tf.math.cosh(0.5 * tau * x)
+    term_2 = ((self.mean_reversion - z2 * tf.math.pow(self.volvol, 2))
+               * tf.math.sinh(0.5 * tau * x)) / x
+
+    return term_1 + term_2
+
+  def F_tilde(self, z1, z2, tau):
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:405]: nit: f_tilde to follow Python style guide

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name='expiries')
+    expiries = tf.broadcast_to(expiries, strikes.shape)
+    sampling_times = tf.convert_to_tensor(
+        sampling_times, dtype=dtype, name='sampling_times')
+    sampling_times = tf.broadcast_to(
+        sampling_times, [sampling_times.shape[0], strikes.shape[0]])
+
+    # Fail assertion if sampling comes after expiry
+    if averaging_frequency == AveragingFrequency.DISCRETE:
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:206]: maybe raise NotImplementedError when  `averaging_frequency` is not discrete?

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    #    - strikes \to strikes / prefactor
+    #    - k_star \to #_past_fixings
+    #    - prepend #_past_fixings negative values to sampling times
+    #    - prices \to prefactor * prices
+    #
+    #  This is slightly different to the approach in the paper, where instead
+    #  t is allowed to increase above 0 as pricing time moves forwards
+
+    if past_fixings is None:
+      running_accumulator = tf.convert_to_tensor(1.0, dtype=dtype)
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:264]: nit: here use tf.constant as 1.0 is a constant

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+    else:
+      running_accumulator = tf.reduce_prod(past_fixings, 0)
+      fixing_count = past_fixings.shape[0]
+
+    start_times = tf.zeros(spots.shape[0], dtype=dtype)
+    tau_k = tf.concat([tf.expand_dims(start_times, axis=0),
+                      sampling_times,
+                      tf.expand_dims(expiries, axis=0)], 0)
+
+    n = sampling_times.shape[0] + fixing_count
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:275]: let's try to support dynamic shapes where possible:

from tf_quant_finance import utils

n = tff.utils.get_shape(sampling_times)[0]

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    prefactor = tf.math.pow(running_accumulator, 1 / n)
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:276]: nit: Prefer overladed operators, where possible:  prefactor = running_accumulator ** (1 / n)

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+    adj_strikes = strikes / prefactor
+
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:278]: Please add a comment to what the shape of `sampling_times` is, e.g., # Shape `[batch_shape, num_times]`.

This is very useful for debugging and also reading through the code

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+    sampling_times = tf.concat(
+      [
+        tf.constant(
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:281]: here sampling_times.shape[1] -> utils.get_shape(sampling_times)[1]. This can be a tensor, so need to replace tf.constant -> tf.convert_to_tensor

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+          [[-1] * sampling_times.shape[1]] * fixing_count,
+          dtype=dtype,
+          shape=(fixing_count, sampling_times.shape[1])
+        ),
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:285]: also, is this the same as 
`-1 * tf.ones([fixing_count, utils.get_shape(sampling_times)[1]],
                      dtype=dtype)`?

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+    z_k_tensor = self.z(s, w)
+    omega_tilde = self.omega_tilde_k(s, w, z_k_tensor)
+    tte = (self.expiries - self.start_times)
+
+    term_1 = self.a(s, w) + omega_tilde[0,:] * self.variances
+    term_2 = (tf.math.pow(self.mean_reversion, 2) * self.theta
+              * tte / tf.math.pow(self.volvol, 2))
+    term_3 = 2 * self.mean_reversion * self.theta / tf.math.pow(self.volvol, 2)
+
+    summation = tf.reduce_sum(tf.math.log(self.F(
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:493]: could you please add a comment with the expected shape?

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+
+  def omega_tilde_k(self, s, w, z_k_tensor):
+    """Calculates \tilde{\omega}_k(s, w) (eq. (19) from the referenced paper)"""
+    omega_tilde = self.omega_k(s, w)
+    mr_v2 = self.mean_reversion / tf.math.pow(self.volvol, 2)
+
+    rows = []
+    last_row = None
+
+    for q in reversed(range(0, omega_tilde.shape[1])):
```
Comment (cyrilchim) [inline tf_quant_finance/models/heston/approximations/asian_prices.py:463]: seems like this can be vecotrized. could you please try using `tf.vecorized_map`  over q here?

**Review by cyrilchim — COMMENTED** [review]
Thank you so much! Mainly a bunch of nits.

File: tf_quant_finance/models/heston/approximations/asian_prices_test.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+    dt = 1. / num_timesteps
+    strikes_tensor = tf.constant(strikes, dtype=dtype)
+    times = tf.constant(np.linspace(0.25, T, 4), dtype=dtype)
+    log_spot = np.log(spots)
+
+    total_fixings = len(past_fixings) + times.shape[0]
+    running_accumulator = np.prod(past_fixings)
+
+    def set_up_pricer(expiries, watch_params=False):
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices_test.py]: Great spot - I've removed both of the wrapping functions, much tidier now

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 3 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:206]: There in lines 189-190 of the latest version

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 4 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Done

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 5 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Fixed

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 6 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Fixed this here.

I also have several expressions of the Heston parameters like `tf.math.pow(self.volvol, 2)` scattered through the code, should these all be replaced with `self.volvol ** 2` as well?

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 7 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:280]: Added

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 8 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Done

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 9 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Great spot, it is exactly. Updated, along with a comment on shape.

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+  ```
+
+  #### References:
+  [1] B. Kim, J. Kim, J. Kim & I. S. Wee, "A Recursive Method for Discretely
+    Monitored Geometric Asian Option Prices", Bull. Korean Math. Soc. 53,
+    733-749 (2016)
+
+  Args:
+    variances: A real `Tensor` of the same dtype and compatible shape as
+      `strikes`. The initial value of the variance.
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Updated

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+    """Calculates a(s, w) (eq. (17) from the referenced paper)"""
+    raw_drift = self.discount_rates - self.dividend_rates
+    corr_drift = self.rho * self.mean_reversion * self.theta / self.volvol
+    mod_drift = raw_drift - corr_drift
+
+    summed_samplings = tf.expand_dims(
+      tf.reduce_sum(self.sampling_times[self.k_star:], axis=0),
+      -1
+    )
+
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Updated

File: tf_quant_finance/models/heston/approximations/asian_prices_test.py [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        initial_state = tf.constant([log_spot, variances], dtype=dtype)
+        paths = process.sample_paths(
+          times,
+          initial_state,
+          time_step=dt,
+          num_samples=num_samples,
+          seed=123
+        )
+
+        prefactor = pow(running_accumulator, 1 / total_fixings)
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices_test.py]: Fixed (but see comment below on the other similar point)

File: tf_quant_finance/models/heston/approximations/asian_prices_test.py [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+        spot_prods = prefactor * tf.math.exp(tf.reduce_sum(paths[:,:,0], 1) / total_fixings)
+        spot_prods = tf.broadcast_to(tf.expand_dims(spot_prods, 1), spot_prods.shape + strikes_tensor.shape)
+        payoffs = tf.nn.relu(spot_prods - strikes_tensor)
+
+        return tf.math.reduce_mean(payoffs, 0)
+      return price_asian_options
+
+    price_asian_options = tf.function(set_up_pricer(expiries))
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices_test.py]: Done

File: tf_quant_finance/models/heston/approximations/asian_prices_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices_test.py]: Great spot, I've removed both of the wrapping functions now and just price directly, much tidier.

[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices_test.py]: Lol github

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 11 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:464]: Ahhh this is a tricky issue. \tilde{\omega} is calculated recursively (it's where the Recursive in the paper's title comes from actually), I've reproduced it here:

image

You can see it's very serial - each row must be calculated and then plugged into the next row's calculation in a very non-linear fashion. I tried lots of tricks to untangle it but haven't been able to so far, in the end this loop formulation was the clearest thing I could settle on.

I'd love to have any thoughts on ways we might manage this better. For example, definitely the final row would need to come after all of the preceding rows have been calculated, but we *could* start any part of the calculation that only depends on the first x rows as soon as those have been calculated - if there are any tricks you can think of here I'd love to see them.

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 1 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:483]: I thought a bit about this and went with `Phi` instead of `phi` in the end because it refers to a function in the reference paper that the authors call \Phi.

Although lowercase functions are preferred in the PEP8 style guide, it also says that for the sake of clarity and explicitness we can occasionally break the rules.

I'm happy to defer if you have very strong feelings on this but I prefer matching the reference here.

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 2 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:407]: As above - this is to match two equations in the reference paper, but happy to defer if you feel very strongly here.

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 11 above)
Comment (cyrilchim) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:464]: ok, this makes sense. It is better to manually add tf.while_loop inside the library as otherwise, when [Autograph](https://www.tensorflow.org/api_docs/python/tf/autograph) is disabled, this may unroll to an unnecessary large graph.

So something similar to [Euler sampling](https://github.com/google/tf-quant-finance/blob/7a8ed52958fc21a168d8fd419a339008f521d846/tf_quant_finance/models/euler_sampling.py#L405):

```
num_iterations = utils.get_shape(omega_tilde)[1]

rows = tf.TensorArray(dtype=dtype,
                            size=num_iterations,
                            element_shape='shape of omega?',
                            clear_after_read=Fals)

def cond(it, omega_tilde, rows):
  return  it < num_iterations

def body(it, omega_tilde, rows):
   ...
  row = ...
  ...
  rows = rows.write(it, row)
  return it + 1, omega_tilde, rows

_, _, rows = tf.while_loop()

rows = rows.stack()
```

Please let me know if you need help filling in the details

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 1 above)
Comment (cyrilchim) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:483]: internally we have to follow the style guide. Could you please rename to `phi` and comment that this is the same as `Phi`?

**Review by cyrilchim — COMMENTED** [review]
Sorry for taking a while to review. There is only one main comment to add a while_loop explicitly. The rest looks very good! Thank you!

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+      **kwargs
+    )
+
+    asian_forward = tf.cast(
+      tf.math.real(asian_prices_handler.Phi(1, 0)),
+      dtype=dtype
+    )
+    dcfs = tf.math.exp(-discount_rates * expiries)
+
+    calls = prefactor * dcfs * ((asian_forward - strikes) / 2
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:364]: Fixed

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+                                + tf.expand_dims(integral, -1) / np.pi)
+
+    if is_call_options is None:
+      return tf.reshape(calls, results_shape)
+
+    puts = prefactor * dcfs * ((strikes - asian_forward) / 2
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py:376]: Fixed

File: tf_quant_finance/models/heston/approximations/asian_prices.py [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+                               + tf.expand_dims(integral, -1) / np.pi)
+    predicate = tf.broadcast_to(is_call_options, tf.shape(calls))
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Nice spot - I've changed this slightly to reflect your feedback, and also added an additional test with an `is_call_options` input

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 1 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Renamed

File: tf_quant_finance/models/heston/approximations/asian_prices.py (hunk 11 above)
Comment (jackgillett101) [reply tf_quant_finance/models/heston/approximations/asian_prices.py]: Done! It works nicely and the tests are passing. Thanks for this feedback, it's much nicer than the original for loop solution.

**Comment by cyrilchim** [comment]
Looks very good! Thank you! Could you please squash and I will start the merge process

**Comment by jackgillett101** [comment]
Done

**Comment by cyrilchim** [comment]
Thank you, @jackgillett101 and so sorry for the wait! I have cleaned up the code a bit to make sure it works with dynamic shapes and have merged the result
